src/visualizers/Chaos.ts

```diff
@@ -1,429 +1,804 @@
+                    )
+                ) ?? -1 // invalid
+            if (this.statusOf.cornerFormula.invalid()) return
+            if (myCorner < 0 || myCorner >= this.corners) {
+                this.statusOf.cornerFormula.warnings.length = 0
+                this.statusOf.cornerFormula.addWarning(
+                    'some cornerFormula values not in corner range'
+                )
+                myCorner = -1 // invalid
+            }
+            if (myCorner == -1) {
+                continue
+            }
+
+            const input = {
+                n: Number(i),
+                a: Number(currElement),
+                s: Number(i - this.seq.first + 1n),
+                m: Number(this.seq.first),
+                M: Number(this.seq.last),
+                f: currentFrames,
+                c: lastCorner,
+                p: this.corners,
+                C: myCorner,
+                w: currWalker,
+                W: this.walkers,
+                x: position.x,
+                y: position.y,
+                A: (n: number | bigint) =>
+                    Number(this.seq.getElement(BigInt(n))),
+            }
+
+            // determine new color
+            let clr: unknown = null
+            clr =
+                this.colorFormula.computeWithStatus(
+                    this.statusOf.colorFormula,
+                    input
+                ) ?? 0
+            if (this.statusOf.colorFormula.invalid()) return
+            // determine new dot size
+            let circSize = 0
+            circSize =
+                math.safeNumber(
+                    this.sizeFormula.computeWithStatus(
+                        this.statusOf.sizeFormula,
+                        input
+                    )
+                ) ?? 1
+            if (this.statusOf.sizeFormula.invalid()) return
+            // determine new stepsize
+            let step = 0
+            step =
+                math.safeNumber(
+                    this.stepFormula.computeWithStatus(
+                        this.statusOf.stepFormula,
+                        input
+                    )
+                ) ?? 0
+            if (this.statusOf.stepFormula.invalid()) return
+            // determine new position
+            const myCornerPosition = this.cornersList[myCorner]
+            position.lerp(myCornerPosition, step)
+
+            // push everything if valid
+            this.dots[currWalker].push(position.copy())
+            this.dotsSizes[currWalker].push(math.safeNumber(circSize))
+            this.dotsIndices[currWalker].push(i)
+            this.dotsCorners[currWalker].push(myCorner)
+            if (typeof clr === 'string') {
+                this.dotsColors[currWalker].push(this.sketch.color(clr))
+            } else if (math.isChroma(clr)) {
+                this.dotsColors[currWalker].push(this.sketch.color(clr.hex()))
+            } else
+                this.dotsColors[currWalker].push(this.sketch.color('white'))
```

Comment:
You could make this whole thing unconditional by just always using `this.sketch.color(chroma(clr).hex())`. That should work regardless of what type/format `clr` is.

Reply:
I can't figure out how.

Reply:
Wow right yes typescript made this a big pain. I found a way to avoid taking cases. This should improve a lot when we have the "desired type" feature for formula parameters, after the mathjs update.